src/extra/widgets/animbtn/lv_animbtn.c

```diff
@@ -0,0 +1,365 @@
+/**
+ * @file lv_animbtn.c
+ *
+ */
+
+/*********************
+ *      INCLUDES
+ *********************/
+
+#include "lv_animbtn.h"
+
+#if LV_USE_ANIMBTN != 0
+
+/*********************
+ *      DEFINES
+ *********************/
+#define MY_CLASS &lv_animbtn_class
+
+/**********************
+ *      TYPEDEFS
+ **********************/
+
+/**********************
+ *  STATIC PROTOTYPES
+ **********************/
+static void lv_animbtn_constructor(const lv_obj_class_t * class_p, lv_obj_t * obj);
+static void lv_animbtn_destructor(const lv_obj_class_t * class_p, lv_obj_t * obj);
+static void lv_animbtn_event(const lv_obj_class_t * class_p, lv_event_t * e);
+static void apply_state(lv_obj_t * animbtn, bool skip_transition);
+static void loop_state(lv_obj_t * animbtn);
+static lv_animbtn_state_t suggest_state(lv_obj_t * animbtn, lv_animbtn_state_t state);
+static lv_animbtn_state_t get_state(const lv_obj_t * animbtn);
+static void setup_anim(lv_animbtn_t * animbtn, lv_animbtn_state_desc_t * desc);
+static int is_state_valid(const lv_animbtn_state_desc_t * state);
+static uint8_t find_trans(lv_animbtn_t * animbtn, lv_animbtn_state_t from, lv_animbtn_state_t to);
+static bool is_transiting(lv_animbtn_t * animbtn, lv_animbtn_state_t current_state);
+
+/**********************
+ *  STATIC VARIABLES
+ **********************/
+const lv_obj_class_t lv_animbtn_class = {
+    .base_class = &lv_img_class,
+    .instance_size = sizeof(lv_animbtn_t),
+    .constructor_cb = lv_animbtn_constructor,
+    .destructor_cb = lv_animbtn_destructor,
+    .event_cb = lv_animbtn_event,
+};
+
+/**********************
+ *      MACROS
+ **********************/
+
+/**********************
+ *   GLOBAL FUNCTIONS
+ **********************/
+
+lv_obj_t * lv_animbtn_create(lv_obj_t * parent)
+{
+    LV_LOG_INFO("begin");
+    lv_obj_t * obj = lv_obj_class_create_obj(MY_CLASS, parent);
+    lv_obj_class_init_obj(obj);
+
+    lv_img_t * img = (lv_img_t *)obj;
```

Comment:
Why do you need to an `lv_obj` wrapper?

Reply:
I'm not really sure how the object inheritance works in LVGL. I want to create an instance of lv_img_t here by calling its constructor and then just fixing the additional members.
So I've set `MYCLASS` to refer as lv_img's class for the base class, but I'm calling `lv_obj_class_create_obj/class_init_obj` here. Is it right ?

I guess the 2 first lines above will call animbutton_constructor (it should since I've specified it in MYCLASS right ?)
But `lv_obj_class_create_obj` returns a `lv_obj_t*` and I need to access the `lv_img_t`'s width and height member here (which might not be those of the object, for example for vector picture). 

Please correct if I'm missing something here.

Reply:
Ping ?

Reply:
Before the last change, one has to pass a fully constructed img object to the create function. So the image's width and height were supposedly set (because of `lv_img_set_src`). The anim button contained an image so the image's size was used to set the button's size.

Now, the image object isn't constructed beforehand, so I must provide a mechanism for this. 
By looking at the image decoder PR, I've removed all image's decoder specific code bu lv_*pic*_init() except for rlottie.

So it's perfectly fine to let the user call `lv_img_set_src` on a `lv_animbtn_t` object. I'll need to add a parameter in the function `lv_img_set_src_ext` to specify a image decoder's context to use while constructing (typically, that's where we set the image's preferred rendering size for vector format, and the image's buffer size and so on).